File: 0062-unique-paths/0062-unique-paths.py
# recursion+memoization - dp on grids
class Solution:
    def recur(self, i,j, dp):
        if i == 0 and j == 0:
            return 1
        if i<0 or j<0:
            return 0
        if dp[i][j] != -1:
            return dp[i][j]
        l = self.recur(i-1, j, dp)
        r = self.recur(i,j-1, dp)
        dp[i][j] = l+r
        return l+r
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1 for i in range(n)] for i in range(m)]
        return self.recur(m-1, n-1, dp)

File: 0062-unique-paths/0062-unique-paths.java
// T- O(n*m)
// S- O(n*m)
// tabulation + dp array
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        dp[0][0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0)
                    dp[i][j] = 1;
                else {
                    int l = 0, r = 0;
                    if (i != 0)
                        l += dp[i - 1][j];
                    if (j != 0)
                        r += dp[i][j - 1];
                    dp[i][j] = l+r;
                }
            }
        }
        return dp[m-1][n-1];
    }
}